### PR TITLE
Refactor envoy package

### PR DESCRIPTION
**Pull request by kylehodgetts**
This PR refactors the envoy package to use sensible struct types with methods that encapsulate the logic of constructing various parts of the envoy configuration without exposing it to the controllers

## Checklist

- [x] tested locally
- [x] added new dependencies
- [ ] updated the docs
- [x] added a test

### CODE REVIEW

**Review by dobegor — APPROVED** [review]

File: envoy/config/config.go [hunk 1]
```diff
@@ -50,77 +49,56 @@ import (
...
+func (e *EnvoyConfiguration) AddRouteToVHost(vhost string, rt *route.Route) error {
+	virtualHost, ok := e.vHosts[vhost]
+
+	if !ok {
+		return fmt.Errorf("envoy configuration doesnt have virtualhost: %s", vhost)
 	}
-	return false
+
+	if err := virtualHost.AddRoute(rt); err != nil {
+		return fmt.Errorf("route %s already exists for vhost %s", rt.GetName(), vhost)
```
Comment (Tarick) [inline envoy/config/config.go:87]: We shouldn't guess the returned error, AddRoute should generate the message.

File: envoy/types/virtualhost.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		},
+	}
+}
+
+func (v *VirtualHost) AddDomain(domain string) {
+	v.Domains = append(v.Domains, domain)
+}
+
+func (v *VirtualHost) AddRoute(r *route.Route) error {
+	routeExists := func(name string, routes []*route.Route) bool {
```
Comment (Tarick) [inline envoy/types/virtualhost.go:27]: This function routeExists var will be allocated each time we enter AddRoute, and it doesn't require the access to the inner AddRoute variables - i.e. is not justified for the closure definition here, better to move it outside as it was before.

File: controllers/parser.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+				if err != nil {
+					return err
+				}
+
+				rt.Action = routeRoute
+			}
+
+			for _, vh := range envoyConfiguration.GetVirtualHosts() {
+				if err := envoyConfiguration.AddRouteToVHost(vh.Name, rt); err != nil {
+					return fmt.Errorf("failure adding redirect route: %w", err)
```
Comment (Tarick) [inline controllers/parser.go:114]: "redirect route" while it is just a generic route now (i.e. could be a redirect or a route)

File: controllers/parser.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+func generateRoutePath(prefix, path string) string {
+	if prefix == "" {
+		return path
+	}
+
+	// Avoids path joins (removes // in e.g. /path//subpath, or //subpath)
+	return fmt.Sprintf(`%s/%s`, strings.TrimSuffix(prefix, "/"), strings.TrimPrefix(path, "/"))
+}
+
+type generateRouteOpts struct {
```
Comment (Tarick) [inline controllers/parser.go:297]: These are only 2 options to pass to the functions, I don't think that we need this struct for this.

**Review by Tarick — CHANGES_REQUESTED** [review]

File: envoy/types/virtualhost.go (hunk 2 above)
Comment (dobegor) [reply envoy/types/virtualhost.go:27]: No it's not. There's actually no such thing as allocating a function in Go - it's just a syntactic sugar. Compiler will either inline this function or make a package-level one with surrounding function's name as a prefix.

**Review by Tarick — APPROVED** [review]

**Pull request merged**